### PR TITLE
HIVE-26043: Use constraint info when creating RexNodes

### PR DESCRIPTION
**Pull request by kasakrisz**
### What changes were proposed in this pull request?
Add nullability to `ColumnInfo` and use it when creating `RexNodes`

### Why are the changes needed?
When TableScan logical operators are created table constraint information is pulled from the metastore and applied the RelDataType associated to the TS operator.
Constraints like not null has an affect the way Hive rewrites subqueries to joins. Example: if the optimizer is aware that the lhs of an in subquery can not be null a simpler rewrite is chosen. See jira for example plans.

### Does this PR introduce _any_ user-facing change?
No, but `explain` command output may be different in older Hive versions.

### How was this patch tested?
```
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestTezTPCDS30TBPerfCliDriver -Dqfile=query45.q -pl itests/qtest -Pitests
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestMiniLlapLocalCliDriver -Dqfile=subquery_in.q -pl itests/qtest -Pitests
```

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java [hunk 1]
```diff
@@ -3119,65 +3123,24 @@ private RelNode genTableLogicalPlan(String tableAlias, QB qb) throws SemanticExc
...
-            }
-          }
-          if (pkPos == -1) {
-            LOG.error("Column for not null constraint definition " + pkCol + " not found");
-            return rowType;
-          }
-          builder.set(pkPos);
-        }
+      if (notNullConstraints != null &&
+              !notNullConstraints.getNotNullConstraints().isEmpty() &&
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:3132]: Do we need the `isEmpty` check?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java [hunk 2]
```diff
@@ -3119,65 +3123,24 @@ private RelNode genTableLogicalPlan(String tableAlias, QB qb) throws SemanticExc
...
-      List<String> fieldNames = new LinkedList<String>();
-      for (RelDataTypeField rdtf : rowType.getFieldList()) {
-        if (bitSet.indexOf(rdtf.getIndex()) != -1) {
-          fieldTypes.add(dtFactory.createTypeWithNullability(rdtf.getType(), false));
-        } else {
-          fieldTypes.add(rdtf.getType());
-        }
-        fieldNames.add(rdtf.getName());
+      if (primaryKeyInfo != null &&
+              !primaryKeyInfo.getColNames().isEmpty() &&
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:3138]: Do we need the `isEmpty` check?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java [hunk 3]
```diff
@@ -3119,65 +3123,24 @@ private RelNode genTableLogicalPlan(String tableAlias, QB qb) throws SemanticExc
...
-        throws HiveException {
-      final NotNullConstraint nnc = tabMetaData.getNotNullConstraint();
-      final PrimaryKeyInfo pkc = tabMetaData.getPrimaryKeyInfo();
-      if ((nnc == null || nnc.getNotNullConstraints().isEmpty()) &&
-          (pkc == null || pkc.getColNames().isEmpty())) {
-        return rowType;
+    private boolean isNullable(String colName, NotNullConstraint notNullConstraints, PrimaryKeyInfo primaryKeyInfo) {
+      if (notNullConstraints == null && primaryKeyInfo == null) {
+        return true;
       }
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:3129]: Do we need this `if` block? If both are null we are going to reach the final `return true` statement.

**Review by zabetak — COMMENTED** [review]
Overall, LGTM. Can we have NOT NULL or PK constraints on partition/virtual columns? If the answer is yes we may need to update a few other places to include the nullability info.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java (hunk 1 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java]: Removed

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java (hunk 2 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java]: Removed

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java (hunk 3 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java]: Removed

**Comment by kasakrisz** [comment]
Partition columns can have not null constraint defined so I extended this patch to handle them. 

Virtual columns can have null values. Ex.: 
```
SELECT ROW__ID FROM non_transactional;
```
will produce `NULLs`.

**Review by zabetak — APPROVED** [review]
LGTM, pending tests!

**Pull request merged**